include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h

```diff
@@ -85,47 +85,26 @@ void
 __serial_merge(const _Rng1& __rng1, const _Rng2& __rng2, _Rng3& __rng3, _Index __start1, _Index __start2,
                const _Index __start3, const std::uint8_t __chunk, const _Index __n1, const _Index __n2, _Compare __comp)
 {
-    if (__start1 >= __n1)
-    {
-        //copying a residual of the second seq
-        const _Index __n = std::min<_Index>(__n2 - __start2, __chunk);
-        for (std::uint8_t __i = 0; __i < __n; ++__i)
-            __rng3[__start3 + __i] = __rng2[__start2 + __i];
-    }
-    else if (__start2 >= __n2)
-    {
-        //copying a residual of the first seq
-        const _Index __n = std::min<_Index>(__n1 - __start1, __chunk);
-        for (std::uint8_t __i = 0; __i < __n; ++__i)
-            __rng3[__start3 + __i] = __rng1[__start1 + __i];
-    }
-    else
+    const _Index __rng1_size = std::min<_Index>(__n1 > __start1 ? __n1 - __start1 : (_Index)0, (_Index)__chunk);
+    const _Index __rng2_size = std::min<_Index>(__n2 > __start2 ? __n2 - __start2 : (_Index)0, (_Index)__chunk);
+    const _Index __rng3_size = std::min<_Index>(__rng1_size + __rng2_size, (_Index)__chunk);
+
+    const _Index __rng1_idx_end = __start1 + __rng1_size;
+    const _Index __rng2_idx_end = __start2 + __rng2_size;
+    const _Index __rng3_idx_end = __start3 + __rng3_size;
+
+    _Index __rng1_idx = __start1;
+    _Index __rng2_idx = __start2;
+
+    for (_Index __rng3_idx = __start3; __rng3_idx < __rng3_idx_end; ++__rng3_idx)
     {
-        for (std::uint8_t __i = 0; __i < __chunk && __start1 < __n1 && __start2 < __n2; ++__i)
-        {
-            const auto& __val1 = __rng1[__start1];
-            const auto& __val2 = __rng2[__start2];
-            if (__comp(__val2, __val1))
-            {
-                __rng3[__start3 + __i] = __val2;
-                if (++__start2 == __n2)
-                {
-                    //copying a residual of the first seq
-                    for (++__i; __i < __chunk && __start1 < __n1; ++__i, ++__start1)
-                        __rng3[__start3 + __i] = __rng1[__start1];
-                }
-            }
-            else
-            {
-                __rng3[__start3 + __i] = __val1;
-                if (++__start1 == __n1)
-                {
-                    //copying a residual of the second seq
-                    for (++__i; __i < __chunk && __start2 < __n2; ++__i, ++__start2)
-                        __rng3[__start3 + __i] = __rng2[__start2];
-                }
-            }
-        }
+        const bool __rng1_idx_less_n1 = __rng1_idx < __rng1_idx_end;
+        const bool __rng2_idx_less_n2 = __rng2_idx < __rng2_idx_end;
+
+        __rng3[__rng3_idx] =
+            __rng1_idx_less_n1 && __rng2_idx_less_n2
+                ? (__comp(__rng2[__rng2_idx], __rng1[__rng1_idx]) ? __rng2[__rng2_idx++] : __rng1[__rng1_idx++])
```

Comment:
Postfix increment leads to a temporary variable creation usually...

Reply:
Post-increment here required from logic side of implementation: we should get value from range and then increment index.  And exactly increment index for that range from which we got value.

Reply:
> Post-increment here required from logic side of implementation: we should get value from range and then increment index. And exactly increment index for that range from which we got value.

Of course I understand the logic here and I am aware of difference between Post-increment  and Pre-increment  operations in C/C++...

I am talking about possible additional temporary variables creation. Usually it is bad style in C/C++ by performance reasons. So, do we have / might have any performance penalty comparing with code w/o temporary variable in given context?

Reply:
@MikeDvorskiy  I suppose the suggestion would then be to convert the ternary operators into a couple `if`/`else` statements where we can group the increment in a separate expression in the if block rather than embedding it in the assignment.
Is that your suggestion?

Reply:
Since `operator[]` here is likely a function call, and the expressions for the arguments of function calls are evaluated before the call, a temporary variable is technically needed. On the other hand, arguments are passed to the function calls either on the stack or in registers (depending on the calling convention), so the place for that temporary variable is likely to exist anyway.

Overall, I would not bother with such micro-optimizations, given that even explicit loop unrolling turned out to have no notable impact.

Reply:
> Is that your suggestion?

Not necessary  `if/else`.... for example: `(++__rng2_idx, __rng2[__rng2_idx-1])`

```
__rng3[__rng3_idx] =
            __rng1_idx_less__n1 && __rng2_idx_less__n2
                ? (__comp(__rng2[__rng2_idx], __rng1[__rng1_idx]) ? (++__rng2_idx, __rng2[__rng2_idx-1]) : (++__rng1_idx, __rng1[__rng1_idx-1]))
                : (__rng1_idx_less__n1 ? (++__rng1_idx, __rng1[__rng1_idx-1]) : (++__rng2_idx, __rng2[__rng2_idx-1]));
```

Reply:
> Overall, I would not bother with such micro-optimizations, given that even explicit loop unrolling turned out to have no notable impact.

I just ask a question about post-increment, nothing more :)
Additionally, I've just compared the asm files with -03. The are identical. 

So, Q.E.D. `we do believe in a compiler.`, This is something like what I expected to hear as an answer, not explanation how serial merge works :)

Reply:
> for example: (++__rng2_idx, __rng2[__rng2_idx-1])
... where `__rng2_idx-1` would be I believe an rvalue similar to the one created by `__rng2_idx++` :)